halo2_proofs/examples/simple-lookup.rs

```diff
@@ -0,0 +1,269 @@
+// |  10 ||       |       |       |         |     10    |     10    |     00    |
+// |  11 ||       |       |       |         |     10    |     11    |     01    |
+// |  12 ||       |       |       |         |     11    |     00    |     11    |
+// |  13 ||       |       |       |         |     11    |     01    |     10    |
+// |  14 ||       |       |       |         |     11    |     10    |     01    |
+// |  15 ||       |       |       |         |     11    |     11    |     00    |
+// |-----||-------|-------|-------|---------|-----------|-----------|-----------|
+//
+// where row 0 is used to allocate the prover's private inputs (`a` and `b`) and public input (`c`).
+// Row 0 is also used to allocate the verifier's public input (`PI`) where consistency between the
+// prover's and verifier's public inputs `c = PI` is checked using a public-input.
+// Every legal input-output tuples for the 2-bit XOR function `xor(l, r) = o` is written
+// into a row of the fixed columns `xor_r_col`, `xor_l_col`, and `xor_o_col`. Note that this circuit
+// does not range check any of `a`, `b`, or `c` to be 2-bits, i.e. there are no polynomial
+// constraints which explicity check `a, b, c ∈ {0, 1, 2, 3}`.
+
+use halo2_proofs::{
+    circuit::{AssignedCell, Chip, Layouter, SimpleFloorPlanner},
+    dev::{MockProver, VerifyFailure},
+    pasta::Fp,
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error, Instance, TableColumn},
+    poly::Rotation,
+};
+
+const XOR_BITS: usize = 2;
+
+struct XorChip {
+    config: XorChipConfig,
+}
+
+#[derive(Clone, Debug)]
+struct XorChipConfig {
+    l_col: Column<Advice>,
+    r_col: Column<Advice>,
+    o_col: Column<Advice>,
+    xor_l_col: TableColumn,
+    xor_r_col: TableColumn,
+    xor_o_col: TableColumn,
+    pub_col: Column<Instance>,
+}
+
+impl Chip<Fp> for XorChip {
+    type Config = XorChipConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl XorChip {
+    fn new(config: XorChipConfig) -> Self {
+        XorChip { config }
+    }
+
+    fn configure(cs: &mut ConstraintSystem<Fp>) -> XorChipConfig {
+        let l_col = cs.advice_column();
+        let r_col = cs.advice_column();
+        let o_col = cs.advice_column();
+        let pub_col = cs.instance_column();
+
+        cs.enable_equality(l_col);
+        cs.enable_equality(r_col);
+        cs.enable_equality(o_col);
+        cs.enable_equality(pub_col);
+
+        let xor_l_col = cs.lookup_table_column();
+        let xor_r_col = cs.lookup_table_column();
+        let xor_o_col = cs.lookup_table_column();
+
+        let _ = cs.lookup(|cs| {
+            vec![
+                (cs.query_advice(l_col, Rotation::cur()), xor_l_col),
+                (cs.query_advice(r_col, Rotation::cur()), xor_r_col),
+                (cs.query_advice(o_col, Rotation::cur()), xor_o_col),
+            ]
+        });
```

Comment:
It could be instructive for this example to gate these lookup arguments on a selector. Currently, it's working without a selector because unassigned cells default to zero, and `(0, 0, 0)` happens to be a row in the lookup table.

This selector should be included in the config and enabled in `alloc_private_and_public_inputs()`.

```suggestion
        let q_lookup = meta.complex_selector();
        let not_q_lookup = Expression::Constant(Fp::one()) - q_lookup.clone();
        
        // Default values to provide to the lookup argument when `q_lookup` is not enabled.
        let (default_l, default_r, default_o) = {
            let one = Expression::Constant(Fp::one());
            let zero = Expression::Constant(Fp::zero());
            (not_q_lookup.clone() * one.clone(), not_q_lookup.clone() * one.clone(), not_q_lookup * zero)
        };
        
        let _ = cs.lookup(|cs| {
            vec![
                (q_lookup.clone() * cs.query_advice(l_col, Rotation::cur()) + default_l, xor_l_col),
                (q_lookup.clone() * cs.query_advice(r_col, Rotation::cur()) + default_r, xor_r_col),
                (q_lookup * cs.query_advice(o_col, Rotation::cur()) + default_o, xor_o_col),
            ]
        });
```